### PR TITLE
[mlir][emitc] Restrict types in EmitC

### PR DESCRIPTION
**Pull request by TinaAMD**
Restrict the types which are valid for EmitC operations. Use what is currently supported by the emitter as restriction. Define a utility functions for valid types, such that they can be used to restrict the operations in table gen as well as being available for reuse in dialect conversions.

### CODE REVIEW

**Review by marbre — APPROVED** [review]
That is a nice improvement, thanks!

**Comment by TinaAMD** [comment]
@simon-camp do you have time to review?

**Review by simon-camp — APPROVED** [review]

**Comment by marbre** [comment]
> @simon-camp do you have time to review?

It is perfectly fine for us if either @simon-camp or me review. So feel free to just go ahead with merging after an approval of Simon or me. No need to wait for a second approval if you feel confident with this.

**Comment by TinaAMD** [comment]
> > @simon-camp do you have time to review?
> 
> It is perfectly fine for us if either @simon-camp or me review. So feel free to just go ahead with merging after an approval of Simon or me. No need to wait for a second approval if you feel confident with this.

Thanks for letting me know, I'll go with that policy then 🙂

**Comment by fifield** [comment]
@TinaAMD Is there a reason `VectorType` and `bf16` need to be excluded here? We use EmitC successfully with both types in [mlir-aie](https://github.com/Xilinx/mlir-aie), and this is a breaking change there.

**Comment by TinaAMD** [comment]
I aligned the restriction to the capabilities of the [Emitter](https://github.com/llvm/llvm-project/blob/15027be6a5c5cf870f25af8de05d6d06c767238f/mlir/lib/Target/Cpp/TranslateToCpp.cpp#L1537), such that one could only build EmitC code which the emitter can handle (to be able to find errors upon construction already). As far as I can see, the emitter doesn't support `VectorType` and `bf16`; do you have downstream changes for these types?

**Comment by fifield** [comment]
> I aligned the restriction to the capabilities of the [Emitter](https://github.com/llvm/llvm-project/blob/15027be6a5c5cf870f25af8de05d6d06c767238f/mlir/lib/Target/Cpp/TranslateToCpp.cpp#L1537), such that one could only build EmitC code which the emitter can handle (to be able to find errors upon construction already). As far as I can see, the emitter doesn't support `VectorType` and `bf16`; do you have downstream changes for these types?

mlir-aie does not use the Emitter directly, but it does have a [similar pass](https://github.com/Xilinx/mlir-aie/blob/main/lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp) supporting vector and bf16 types, and that pass reuses ops from emitc dialect. Because those ops are mixed with "unsupported" types, the emitc verifier fails.

To me it seems like the legality check for the Emitter should be in the Emitter, not in the dialect. Then the input to the Emitter can be generated by a multi-stage lowering, where the intermediate IR(s) might contain things not supported by the Emitter, but those things will be legalized away by subsequent transformations.

**Comment by mgehre-amd** [comment]
> To me it seems like the legality check for the Emitter should be in the Emitter, not in the dialect. Then the input to the Emitter can be generated by a multi-stage lowering, where the intermediate IR(s) might contain things not supported by the Emitter, but those things will be legalized away by subsequent transformations.
I prefer the current design because the emitc dialects is exactly meant to represent what can be emitted into C.
For user defined types, like `bf16`, it allows to represent them as a `!emitc.opaque<"mybfloat16_t">` type, which then also doesn't require to change the emitter.

fyi @marbre

**Comment by fifield** [comment]
Thanks all for answering my questions and providing the possible workaround.

**Comment by jsetoain** [comment]
> We have had these discussions several times online and offline (already during upstreaming the dialect and the emitter), and the consensus was to make the dialect robust and keep the emitter simple. As @mgehre-amd mentioned, types that are not supported by the emitter can be represented with `!emitc.opaque<"mytype">`. Of course we are also happy to discuss if it makes sense to add support for further types to the dialect / emitter.

Just to clarify, does this mean that the right way to have a call to a function with unsupported types is to first `emit.cast` them into opaque types? What's breaking for us right now is an `emitc.opaque_call` to a function that takes bf16 vectors as parameters, and I'm trying to figure out what's the right thing to do here 😅

**Comment by mgehre-amd** [comment]
> > We have had these discussions several times online and offline (already during upstreaming the dialect and the emitter), and the consensus was to make the dialect robust and keep the emitter simple. As @mgehre-amd mentioned, types that are not supported by the emitter can be represented with `!emitc.opaque<"mytype">`. Of course we are also happy to discuss if it makes sense to add support for further types to the dialect / emitter.
> 
> Just to clarify, does this mean that the right way to have a call to a function with unsupported types is to first `emit.cast` them into opaque types? What's breaking for us right now is an `emitc.opaque_call` to a function that takes bf16 vectors as parameters, and I'm trying to figure out what's the right thing to do here 😅
> 
> Edit: I've tried the casting approach, and the only thing that seems to be working for me is injecting `unrealized_conversion_cast`, which is an issue because it won't be reconciled in any natural way. Am I right to understand that the use we're abusing `emitc` with the use we're making of it? Is there a nicer way to interface with non-emitc types?

emitc is about emitting C code, so the first question becomes: How should a `vector of bf16` be spelled in the C code? You then put that spelling into a `!emitc.opaque`, e.g. ``!emitc.opaque<"myvector<mybf16>">`. And then you have to use that everywhere in the emitc IR (which typically means that you will use a `TypeConverter` during dialect conversion).
You cannot use `emitc.cast` to convert an illegal type into an `!emitc.opaque`, because then the `cast` couldn't be emitted as C code.

**Comment by TinaAMD** [comment]
Assuming you would expect a bfloat16 represented as follows in your C code:

```
reprOfYourBf16Type myVar; // Variable declaration of myVar with the bfloat16 type
```
When you convert from some dialect that uses bfloat16 to EmitC, you can add a [type converter](https://mlir.llvm.org/docs/DialectConversion/#type-conversion) that describes what your type will look like in EmitC.
Converting bfloat16 could look similar to this:

```
TypeConverter typeConverter;
typeConverter.addConversion([this](Type type) -> std::optional<Type> {
  // Convert bfloat16 to emitc.opaque
  if (type.isBF16()) {
    return emitc::OpaqueType::get(this->getContext(), "reprOfYourBf16Type");
  }
  // ...possibly more conversions....
  return type;  // default for all types which are allowed in EmitC
});
```

**Comment by jsetoain** [comment]
> Assuming you would expect a bfloat16 represented as follows in your C code:
> 
> ```
> reprOfYourBf16Type myVar; // Variable declaration of myVar with the bfloat16 type
> ```
> 
> When you convert from some dialect that uses bfloat16 to EmitC, you can add a [type converter](https://mlir.llvm.org/docs/DialectConversion/#type-conversion) that describes what your type will look like in EmitC. Converting bfloat16 could look similar to this:
> 
> ```
> TypeConverter typeConverter;
> typeConverter.addConversion([this](Type type) -> std::optional<Type> {
>   // Convert bfloat16 to emitc.opaque
>   if (type.isBF16()) {
>     return emitc::OpaqueType::get(this->getContext(), "reprOfYourBf16Type");
>   }
>   // ...possibly more conversions....
>   return type;  // default for all types which are allowed in EmitC
> });
> ```

I see. I did understand, then. I believe we're indeed abusing the dialect because we don't convert everything to emitc before translating, we end up with a mix of dialects before translation so we end up with unreconciled unrealized casts. I will work around this the best I can, thank you both for your clarifications!

**Pull request merged**
